src/http/server.rs

```diff
@@ -21,38 +22,76 @@ fn graphiql() -> Result<Response<Body>> {
   )))
 }
 
-pub async fn graphql_request(req: Request<Body>, server_ctx: &ServerContext) -> Result<Response<Body>> {
+fn not_found() -> Result<Response<Body>> {
+  Ok(Response::builder().status(StatusCode::NOT_FOUND).body(Body::empty())?)
+}
+
+fn create_request_context(req: &Request<Body>, server_ctx: &ServerContext) -> RequestContext {
   let upstream = server_ctx.blueprint.upstream.clone();
   let allowed = upstream.get_allowed_headers();
   let headers = create_allowed_headers(req.headers(), &allowed);
-  let bytes = hyper::body::to_bytes(req.into_body()).await?;
-  let request: async_graphql_hyper::GraphQLRequest = serde_json::from_slice(&bytes)?;
-  let req_ctx = Arc::new(RequestContext::from(server_ctx).req_headers(headers));
-  let mut response = request.data(req_ctx.clone()).execute(&server_ctx.schema).await;
+  RequestContext::from(server_ctx).req_headers(headers)
+}
+
+fn update_cache_control_header(
+  response: GraphQLResponse,
+  server_ctx: &ServerContext,
+  req_ctx: Arc<RequestContext>,
+) -> GraphQLResponse {
   if server_ctx.blueprint.server.enable_cache_control_header {
     if let Some(ttl) = req_ctx.get_min_max_age() {
-      response = response.set_cache_control(ttl as i32);
+      return response.set_cache_control(ttl as i32);
     }
   }
-  let mut resp = response.to_response()?;
+  response
+}
+
+pub fn update_response_headers(resp: &mut hyper::Response<hyper::Body>, server_ctx: &ServerContext) {
   if !server_ctx.blueprint.server.response_headers.is_empty() {
     resp
       .headers_mut()
       .extend(server_ctx.blueprint.server.response_headers.clone());
   }
+}
 
+pub async fn graphql_single_request(req: Request<Body>, server_ctx: &ServerContext) -> Result<Response<Body>> {
+  let req_ctx = Arc::new(create_request_context(&req, server_ctx));
+  let bytes = hyper::body::to_bytes(req.into_body()).await?;
+  let request: async_graphql_hyper::GraphQLRequest = serde_json::from_slice(&bytes)?;
+  let mut response = request.data(req_ctx.clone()).execute(&server_ctx.schema).await;
+  response = update_cache_control_header(response, server_ctx, req_ctx);
+  let mut resp = response.to_response()?;
+  update_response_headers(&mut resp, server_ctx);
   Ok(resp)
 }
-fn not_found() -> Result<Response<Body>> {
-  Ok(Response::builder().status(StatusCode::NOT_FOUND).body(Body::empty())?)
+
+pub async fn graphql_batch_request(req: Request<Body>, server_ctx: &ServerContext) -> Result<Response<Body>> {
+  let req_ctx = Arc::new(create_request_context(&req, server_ctx));
+  let bytes = hyper::body::to_bytes(req.into_body()).await?;
+  let request: async_graphql_hyper::GraphQLBatchRequest = serde_json::from_slice(&bytes)?;
+  let mut response = request.data(req_ctx.clone()).execute(&server_ctx.schema).await;
+  response = update_cache_control_header(response, server_ctx, req_ctx);
+  let mut resp = response.to_response()?;
+  update_response_headers(&mut resp, server_ctx);
+  Ok(resp)
 }
```

Comment:
There are two considerations:
1. the decision of which function to use should be made after the config file is read, and not per request
2. We're trying to avoid dynamic dispatch (in my earlier code I had a `dyn` `Fn` which was doing what you suggested)
The difficulty I faced in making it one function is that if the appropriate struct or function needs to be passed in (based on the enable_batch_requests flag, then Rust insists that it should be a `dyn`

Reply:
I see, yeah sometimes it could be quite tricky.

I'm thinking about this way:
```rust
pub async fn graphql_request<T: DeserializeOwned + Data>(req: Request<Body>, server_ctx: &ServerContext) -> Result<Response<Body>> {
  let req_ctx = Arc::new(create_request_context(&req, server_ctx));
  let bytes = hyper::body::to_bytes(req.into_body()).await?;
  let request: T = serde_json::from_slice(&bytes)?;
  let mut response = request.data(req_ctx.clone()).execute(&server_ctx.schema).await;
  response = update_cache_control_header(response, server_ctx, req_ctx);
  let mut resp = response.to_response()?;
  update_response_headers(&mut resp, server_ctx);
  Ok(resp)
}

pub async fn graphql_single_request(req: Request<Body>, server_ctx: &ServerContext) -> Result<Response<Body>> {
  graphql_request::<async_graphql_hyper::GraphQLRequest>(req, server_ctx).await
}

pub async fn graphql_batch_request(req: Request<Body>, server_ctx: &ServerContext) -> Result<Response<Body>> {
  graphql_request::<async_graphql_hyper::GraphQLBatchRequest>(req, server_ctx).await
}
```

So, we still have two different functions without dyn and they should compile to different code due to monomorphization